packages/vite/src/node/optimizer/scan.ts

```diff
@@ -315,21 +315,32 @@ function esbuildScanPlugin(
         let ext = path.extname(id).slice(1)
         if (ext === 'mjs') ext = 'js'
 
+        let loader = ext as Loader
+
+        // Use the user given loader if it is explicitly set
+        if (config.esbuild && config.esbuild.loader) {
+          loader = config.esbuild.loader
+        }
+
         let contents = fs.readFileSync(id, 'utf-8')
-        if (ext.endsWith('x') && config.esbuild && config.esbuild.jsxInject) {
+        if (
+          loader.endsWith('x') &&
+          config.esbuild &&
```

Comment:
**suggestion (if-minor):** Same here `config.esbuild?.jsxInject`

(_No suggested change here cause of multiline_)

Reply:
Same as above